src/controller/CHIPCluster.h

```diff
@@ -205,6 +206,56 @@ class DLL_EXPORT ClusterBase
                                                         mEndpoint, clusterId, attributeId, onSuccessCb, onFailureCb);
     }
 
+    /**
+     * Subscribe to attribute and get a type-safe callback with the attribute
+     * value when it changes.
+     */
+    template <typename AttributeInfo>
+    CHIP_ERROR SubscribeAttribute(void * context, ReadResponseSuccessCallback<typename AttributeInfo::DecodableArgType> reportCb,
+                                  ReadResponseFailureCallback failureCb, uint16_t minIntervalFloorSeconds,
+                                  uint16_t maxIntervalCeilingSeconds,
+                                  SubscriptionEstablishedCallback subscriptionEstablishedCb = nullptr)
+    {
+        return SubscribeAttribute<typename AttributeInfo::DecodableType, typename AttributeInfo::DecodableArgType>(
+            context, AttributeInfo::GetClusterId(), AttributeInfo::GetAttributeId(), reportCb, failureCb, minIntervalFloorSeconds,
+            maxIntervalCeilingSeconds, subscriptionEstablishedCb);
+    }
+
+    template <typename DecodableType, typename DecodableArgType>
+    CHIP_ERROR SubscribeAttribute(void * context, ClusterId clusterId, AttributeId attributeId,
+                                  ReadResponseSuccessCallback<DecodableArgType> reportCb, ReadResponseFailureCallback failureCb,
+                                  uint16_t minIntervalFloorSeconds, uint16_t maxIntervalCeilingSeconds,
+                                  SubscriptionEstablishedCallback subscriptionEstablishedCb = nullptr)
```

Comment:
How do you terminate a subscription after it's been established with such an API? If you were to repeatedly call this with new subscription requests, would that not leak ReadClient's to the point that subscriptions could not be further established?

Reply:
> How do you terminate a subscription after it's been established with such an API?

You don't, just like the API it's a replacement for.

The best you can do is create a new subscription to the same server, which kills the first one on the server and then eventually that gets notified and the ReadClient shut down.

That said, we might in fact want to change this API to hand out the ReadClient so you can shut down the subscription.  We want to get our ReadClient lifetime/handle story straight first, right?